Another gcc 11.1 issue

Not exercised by this library, but by samtools instead, kseq.h also trips up the new gcc 11.1 compiler.

```
gcc11 -g -Wall -O2 -I. -I../htslib -I./lz4  -c -o dict.o dict.c
In file included from dict.c:32:
In function 'kseq_read',
    inlined from 'write_dict' at dict.c:73:17,
    inlined from 'dict_main' at dict.c:176:5:
../htslib/htslib/kseq.h:212:40: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  212 |                 seq->seq.s[seq->seq.l] = 0;     /* null terminated string */ \
      |                 ~~~~~~~~~~~~~~~~~~~~~~~^~~
../htslib/htslib/kseq.h:237:9: note: in expansion of macro '__KSEQ_READ'
  237 |         __KSEQ_READ(SCOPE)
      |         ^~~~~~~~~~~
../htslib/htslib/kseq.h:239:35: note: in expansion of macro 'KSEQ_INIT2'
  239 | #define KSEQ_INIT(type_t, __read) KSEQ_INIT2(static, type_t, __read)
      |                                   ^~~~~~~~~~
dict.c:35:1: note: in expansion of macro 'KSEQ_INIT'
   35 | KSEQ_INIT(gzFile, gzread)
      | ^~~~~~~~~
dict.c: In function 'dict_main':
../htslib/htslib/kseq.h:210:45: note: at offset -2 into destination object of size 0 allocated by 'realloc'
  210 |                         seq->seq.s = (char*)realloc(seq->seq.s, seq->seq.m); \
      |                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../htslib/htslib/kseq.h:237:9: note: in expansion of macro '__KSEQ_READ'
  237 |         __KSEQ_READ(SCOPE)
      |         ^~~~~~~~~~~
../htslib/htslib/kseq.h:239:35: note: in expansion of macro 'KSEQ_INIT2'
  239 | #define KSEQ_INIT(type_t, __read) KSEQ_INIT2(static, type_t, __read)
      |                                   ^~~~~~~~~~
dict.c:35:1: note: in expansion of macro 'KSEQ_INIT'
   35 | KSEQ_INIT(gzFile, gzread)
      | ^~~~~~~~~
```

This ugly mess of macro warnings (which it does a reasonable job of untangling actually), boils down to it not knowing that seq.l < seq.m so it thinks it can overflow the buffer.

It's possibly due to the use of `kroundup32` confusing it.  It doesn't know what's happening there, even if we explicitly use the kroundup.h copy to guarantee it doesn't decrease or overflow.

A trivial bit of compiler education is to put a code path in that handles this case and die.  If this code is ever reached, something terrible has gone wrong, such as the `seq->seq.l + 2` wrapping around.  Frankly once we're doing that we're way outside the parameters of what sensible behaviour is already and crashing and burning is probably appropriate (at least that doesn't constitute buffer overruns or nasty security issues).

What do we feel about something like this?  (I was tempted by abort() given it would have gone bang with a core dump anyway...)

```
diff --git a/htslib/kseq.h b/htslib/kseq.h
index ee13dd0..307486f 100644
--- a/htslib/kseq.h
+++ b/htslib/kseq.h
@@ -207,6 +207,7 @@
                if (seq->seq.l + 1 >= seq->seq.m) { /* seq->seq.s[seq->seq.l] below may be out of boundary */ \
                        seq->seq.m = seq->seq.l + 2; \
                        kroundup32(seq->seq.m); /* rounded to the next closest 2^k */ \
+                       if (seq->seq.l < seq->seq.m) return -3; \
                        seq->seq.s = (char*)realloc(seq->seq.s, seq->seq.m); \
                } \
                seq->seq.s[seq->seq.l] = 0;     /* null terminated string */ \
```

The entire code is missing far more common causes of errors, like running out of memory, but that's a topic for another day.

--- a/htslib/kseq.h
+++ b/htslib/kseq.h
@@ -207,6 +207,7 @@
 		if (seq->seq.l + 1 >= seq->seq.m) { /* seq->seq.s[seq->seq.l] below may be out of boundary */ \
 			seq->seq.m = seq->seq.l + 2; \
 			kroundup32(seq->seq.m); /* rounded to the next closest 2^k */ \
+			if (seq->seq.l + 1 >= seq->seq.m) return -3; /* error: adjusting m overflowed */ \
 			seq->seq.s = (char*)realloc(seq->seq.s, seq->seq.m); \
 		} \
 		seq->seq.s[seq->seq.l] = 0;	/* null terminated string */ \

